tests/unittest/failover3/main_test.go

```diff
@@ -0,0 +1,231 @@
+			continue
+		} else {
+			conn.Close()
+			break
+		}
+	} // */
+
+	os.Exit(testutil.UtilMain(m, cfg, before))
+}
+
+func doCrud(conn *sql.Conn, id int, t* testing.T) (bool) {
+	note := time.Now().Format("test note 2006-01-02j15:04:05.000 failover")
+
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
+
+	stmt, err := conn.PrepareContext(ctx, "create table test_failover ( id int, note varchar(55) )")
+	if err != nil {
+		return false
+	}
+	stmt.Exec()
+	// ignore errors since table may already exist
+
+	// not using txn since mysql
+	stmt, err = conn.PrepareContext(ctx, "insert into test_failover ( id , note ) values ( ?, ? )")
+	if err != nil {
+		t.Fatalf("Error preparing test (insert table) %s\n", err.Error())
+	}
+	_, err = stmt.Exec(id, note)
+	if err != nil {
+		t.Fatalf("Error exec test (insert table) %s\n", err.Error())
+	}
+
+	stmt, err = conn.PrepareContext(ctx, "select note from test_failover where id = ?")
+	if err != nil {
+		t.Fatalf("Error preparing test (sel table) %s\n", err.Error())
+	}
+        rows, _ := stmt.Query(id)
+        if !rows.Next() {
+                t.Fatalf("Expected 1 row")
+        }
+        var str_val sql.NullString
+        err = rows.Scan(&str_val)
+	if err != nil {
+		t.Fatalf("Error preparing test (sel scan table) %s\n", err.Error())
+	}
+	if !str_val.Valid {
+		t.Fatalf("null str")
+	}
+	if str_val.String != note {
+		t.Fatalf("data corrupt "+note+" dbHas:"+ str_val.String)
+	}
+
+        rows.Close()
+        stmt.Close()
+
+
+	stmt, err = conn.PrepareContext(ctx, "delete from test_failover where id = ?")
+	if err != nil {
+		t.Fatalf("Error preparing test (del table) %s\n", err.Error())
+	}
+	_, err = stmt.Exec(id)
+	if err != nil {
+		t.Fatalf("Error preparing test (del table) %s\n", err.Error())
+	}
+	return true
+}
+
+func TestFailover2(t *testing.T) {
+	logger.GetLogger().Log(logger.Debug, "TestFailover2 begin +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
+
+	shard := 0
+	db, err := sql.Open("heraloop", fmt.Sprintf("%d:0:0", shard))
+	if err != nil {
+		t.Fatal("Error starting Mux:", err)
+		return
+	}
+	db.SetMaxIdleConns(0)
+	defer db.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	conn, err := db.Conn(ctx)
+	if err != nil {
+		t.Fatalf("Error getting connection %s\n", err.Error())
+	}
+	defer conn.Close()
+
+	doCrud(conn, 1, t)
+	doCrud(conn, 1, t)
+	conn, err = db.Conn(ctx)
+	if err != nil {
+		logger.GetLogger().Log(logger.Debug, "reacq conn "+err.Error())
+	}
+	doCrud(conn, 1, t)
+
+	logger.GetLogger().Log(logger.Debug, "TestFailover2 taking out first db +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
+
+	mysqlDirect("set global read_only = 1", t)
+
+	logger.GetLogger().Log(logger.Debug, "TestFailover2 taken out first db +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n")
+	//conn.Close() // ?
```

Comment:
Is the connection being held until commit/rollback?  I think we should be able to use it before and after the failover--as long as it hadn't been dedicated.